File: src/main/python/com/example/tutorial/dsa/medium/concurrency/thread_pool_executor_demo.py
"""
ThreadPoolExecutorDemo
----------------------------------
This program demonstrates ThreadPoolExecutor usage.
Shows various thread pool configurations and task submission patterns.

Problem Statement:
    Efficiently manage thread resources for executing multiple tasks concurrently
    without creating excessive threads.

Real UseCase:
    In a credit card offers system:
    - Batch processing of transactions
    - Parallel offer eligibility checks
    - Concurrent API calls to partner services

Company Tags: Amazon, Google, Facebook, Netflix
"""

import sys
import os
import time
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor, as_completed, wait
from typing import List

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers


def task_runner(task_id: int) -> str:
    """Simple task that returns a result."""
    time.sleep(0.1)
    return f"Task-{task_id} completed"


def process_offer(offer_id: int) -> str:
    """Simulates processing an offer."""
    time.sleep(0.1)
    return f"Offer-{offer_id} processed"


def demo_basic_thread_pool():
    """Demonstrates basic ThreadPoolExecutor usage."""
    print("  Submitting 6 tasks to pool of 3 workers:")

    with ThreadPoolExecutor(max_workers=3) as executor:
        futures = []
        for i in range(1, 7):
            future = executor.submit(task_runner, i)
            futures.append(future)

        # Get results as they complete
        for future in as_completed(futures):
            result = future.result()
            print(f"    {result}")

    print("  Basic thread pool demo complete.")


def demo_map_function():
    """Demonstrates executor.map() for parallel processing."""
    print("  Using map() for parallel processing:")

    with ThreadPoolExecutor(max_workers=3) as executor:
        offer_ids = [1, 2, 3, 4, 5]
        results = list(executor.map(process_offer, offer_ids))

        for result in results:
            print(f"    {result}")

    print("  Map function demo complete.")


def demo_context_manager():
    """Demonstrates context manager usage."""
    print("  Using context manager (auto shutdown):")

    with ThreadPoolExecutor(max_workers=2, thread_name_prefix="Worker") as executor:
        future1 = executor.submit(lambda: "Task A done")
        future2 = executor.submit(lambda: "Task B done")

        print(f"    {future1.result()}")
        print(f"    {future2.result()}")

    print("  Context manager demo complete.")


def demo_future_callbacks():
    """Demonstrates Future callbacks."""
    print("  Using Future callbacks:")

    def callback(future):
        print(f"    Callback received: {future.result()}")

    with ThreadPoolExecutor(max_workers=2) as executor:
        future = executor.submit(task_runner, 1)
        future.add_done_callback(callback)

        # Submit another task
        future2 = executor.submit(task_runner, 2)
        future2.add_done_callback(callback)

        # Wait for all to complete
        wait([future, future2])

    print("  Future callbacks demo complete.")


def demo_timeout_handling():
    """Demonstrates timeout handling with futures."""
    print("  Demonstrating timeout handling:")

    def slow_task():
        time.sleep(2)
        return "Slow task done"

    with ThreadPoolExecutor(max_workers=1) as executor:
        future = executor.submit(slow_task)

        try:
            result = future.result(timeout=0.5)
            print(f"    Result: {result}")
        except TimeoutError:
            print("    Task timed out (expected)")
            future.cancel()

    print("  Timeout handling demo complete.")


def demo_exception_handling():
    """Demonstrates exception handling in thread pool."""
    print("  Demonstrating exception handling:")

    def failing_task(task_id: int) -> str:
        if task_id == 2:
            raise ValueError(f"Task {task_id} failed!")
        return f"Task {task_id} succeeded"

    with ThreadPoolExecutor(max_workers=2) as executor:
        futures = {executor.submit(failing_task, i): i for i in range(1, 4)}

        for future in as_completed(futures):
            task_id = futures[future]
            try:
                result = future.result()
                print(f"    {result}")
            except Exception as e:
                print(f"    Task {task_id} raised: {e}")

    print("  Exception handling demo complete.")


def demo_batch_processing():
    """Demonstrates batch processing pattern."""
    print("  Batch processing 10 items in batches of 3:")

    items = list(range(1, 11))
    batch_size = 3

    def process_item(item: int) -> str:
        time.sleep(0.05)
        return f"Processed item {item}"

    with ThreadPoolExecutor(max_workers=3) as executor:
        for i in range(0, len(items), batch_size):
            batch = items[i:i + batch_size]
            print(f"    Processing batch: {batch}")
            results = list(executor.map(process_item, batch))
            for result in results:
                print(f"      {result}")

    print("  Batch processing demo complete.")


def main():
    """Main function to demonstrate the ThreadPoolExecutorDemo."""
    print("=== ThreadPoolExecutorDemo: Thread Pool Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demo 1: Basic Thread Pool
    print("--- Demo 1: Basic Thread Pool ---\n")
    demo_basic_thread_pool()

    # Demo 2: Map Function
    print("\n--- Demo 2: Map Function ---\n")
    demo_map_function()

    # Demo 3: Context Manager
    print("\n--- Demo 3: Context Manager ---\n")
    demo_context_manager()

    # Demo 4: Future Callbacks
    print("\n--- Demo 4: Future Callbacks ---\n")
    demo_future_callbacks()

    # Demo 5: Timeout Handling
    print("\n--- Demo 5: Timeout Handling ---\n")
    demo_timeout_handling()

    # Demo 6: Exception Handling
    print("\n--- Demo 6: Exception Handling ---\n")
    demo_exception_handling()

    # Demo 7: Batch Processing
    print("\n--- Demo 7: Batch Processing ---\n")
    demo_batch_processing()


if __name__ == "__main__":
    main()

File: src/main/java/com/example/tutorial/dsa/medium/concurrency/ThreadPoolExecutorDemo.java
package com.example.tutorial.dsa.medium.concurrency;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

/**
 * ThreadPoolExecutorDemo
 * ----------------------------------
 * <p>This program demonstrates ThreadPoolExecutor and ExecutorService usage.
 * Shows various thread pool configurations and task submission patterns.
 *
 * <p><b>Problem Statement:</b>
 * Efficiently manage thread resources for executing multiple tasks concurrently
 * without creating excessive threads.
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Batch processing of transactions</li>
 *   <li>Parallel offer eligibility checks</li>
 *   <li>Concurrent API calls to partner services</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Amazon, Google, Facebook, Netflix
 */
@Component
public class ThreadPoolExecutorDemo implements CommandLineRunner {
  public static void main(String[] args) {
    SpringApplication.run(ThreadPoolExecutorDemo.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== ThreadPoolExecutorDemo: Thread Pool Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demo 1: Fixed Thread Pool
    System.out.println("--- Demo 1: Fixed Thread Pool ---\n");
    demoFixedThreadPool();

    // Demo 2: Cached Thread Pool
    System.out.println("\n--- Demo 2: Cached Thread Pool ---\n");
    demoCachedThreadPool();

    // Demo 3: Scheduled Thread Pool
    System.out.println("\n--- Demo 3: Scheduled Thread Pool ---\n");
    demoScheduledThreadPool();

    // Demo 4: Custom ThreadPoolExecutor
    System.out.println("\n--- Demo 4: Custom ThreadPoolExecutor ---\n");
    demoCustomThreadPool();

    // Demo 5: Future and Callable
    System.out.println("\n--- Demo 5: Future and Callable ---\n");
    demoFutureCallable();
  }

  private void demoFixedThreadPool() throws InterruptedException {
    ExecutorService executor = Executors.newFixedThreadPool(3);

    System.out.println("  Submitting 6 tasks to pool of 3 threads:");

    for (int i = 1; i <= 6; i++) {
      final int taskId = i;
      executor.submit(() -> {
        System.out.println("    Task-" + taskId + " running on " + Thread.currentThread().getName());
        try {
          Thread.sleep(100);
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
        }
        System.out.println("    Task-" + taskId + " completed");
      });
    }

    executor.shutdown();
    executor.awaitTermination(5, TimeUnit.SECONDS);
    System.out.println("  Fixed thread pool demo complete.");
  }

  private void demoCachedThreadPool() throws InterruptedException {
    ExecutorService executor = Executors.newCachedThreadPool();

    System.out.println("  Submitting 5 tasks to cached pool (creates threads as needed):");

    for (int i = 1; i <= 5; i++) {
      final int taskId = i;
      executor.submit(() -> {
        System.out.println("    Task-" + taskId + " on " + Thread.currentThread().getName());
        try {
          Thread.sleep(50);
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
        }
      });
    }

    executor.shutdown();
    executor.awaitTermination(5, TimeUnit.SECONDS);
    System.out.println("  Cached thread pool demo complete.");
  }

  private void demoScheduledThreadPool() throws InterruptedException {
    ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(2);

    System.out.println("  Scheduling tasks:");

    // One-time delayed execution
    scheduler.schedule(() -> {
      System.out.println("    Delayed task executed after 200ms");
    }, 200, TimeUnit.MILLISECONDS);

    // Fixed rate execution
    ScheduledFuture<?> periodicTask = scheduler.scheduleAtFixedRate(() -> {
      System.out.println("    Periodic task at " + System.currentTimeMillis() % 10000);
    }, 100, 150, TimeUnit.MILLISECONDS);

    // Let it run for a bit
    Thread.sleep(600);
    periodicTask.cancel(false);

    scheduler.shutdown();
    scheduler.awaitTermination(2, TimeUnit.SECONDS);
    System.out.println("  Scheduled thread pool demo complete.");
  }

  /**
   * Demonstrates custom ThreadPoolExecutor configuration.
   *
   * <p><b>Key Parameters:</b>
   * <ul>
   *   <li>corePoolSize: Minimum threads to keep alive</li>
   *   <li>maximumPoolSize: Maximum threads allowed</li>
   *   <li>keepAliveTime: Idle timeout for non-core threads</li>
   *   <li>workQueue: Queue for holding tasks before execution</li>
   *   <li>handler: Policy when queue and threads are full</li>
   * </ul>
   */
  private void demoCustomThreadPool() throws InterruptedException {
    ThreadPoolExecutor executor = new ThreadPoolExecutor(
        2,                      // Core pool size
        4,                      // Max pool size
        60L, TimeUnit.SECONDS,  // Keep alive
        new ArrayBlockingQueue<>(2),  // Bounded queue
        new ThreadPoolExecutor.CallerRunsPolicy()  // Rejection policy
    );

    System.out.println("  Custom pool: core=2, max=4, queue=2");

    for (int i = 1; i <= 8; i++) {
      final int taskId = i;
      System.out.println("  Submitting Task-" + taskId);
      executor.submit(() -> {
        System.out.println("    Task-" + taskId + " running on " + Thread.currentThread().getName());
        try {
          Thread.sleep(100);
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
        }
      });
      Thread.sleep(30);  // Stagger submissions
    }

    executor.shutdown();
    executor.awaitTermination(5, TimeUnit.SECONDS);

    System.out.println("  Custom thread pool demo complete.");
  }

  private void demoFutureCallable() throws InterruptedException, ExecutionException {
    ExecutorService executor = Executors.newFixedThreadPool(2);

    System.out.println("  Submitting Callable tasks that return values:");

    // Submit tasks that return values
    List<Future<String>> futures = new ArrayList<>();

    for (int i = 1; i <= 3; i++) {
      final int offerId = i;
      Future<String> future = executor.submit(() -> {
        Thread.sleep(100);
        return "Offer-" + offerId + " processed";
      });
      futures.add(future);
    }

    // Get results
    for (Future<String> future : futures) {
      String result = future.get();  // Blocks until result available.
      System.out.println("    Result: " + result);
    }

    // InvokeAll - submit all and wait
    System.out.println("\n  Using invokeAll (wait for all):");
    List<Callable<Integer>> tasks = List.of(
        () -> { Thread.sleep(100); return 10; },
        () -> { Thread.sleep(50); return 20; },
        () -> { Thread.sleep(75); return 30; }
    );

    List<Future<Integer>> results = executor.invokeAll(tasks);
    int sum = 0;
    for (Future<Integer> result : results) {
      sum += result.get();
    }
    System.out.println("    Sum of results: " + sum);

    executor.shutdown();
    System.out.println("  Future/Callable demo complete.");
  }
}
